pkg/execution/state/redis_state/queue.go

```diff
@@ -2738,6 +2782,357 @@ func (q *queue) Requeue(ctx context.Context, queueShard QueueShard, i osqueue.Qu
+	}
+}
+
+func (q *queue) BacklogRefill(ctx context.Context, b *QueueBacklog, sp *QueueShadowPartition) error {
+	ctx = redis_telemetry.WithScope(redis_telemetry.WithOpName(ctx, "BacklogRefill"), redis_telemetry.ScopeQueue)
+
+	if q.primaryQueueShard.Kind != string(enums.QueueShardKindRedis) {
+		return fmt.Errorf("unsupported queue shard kind for BacklogRefill: %s", q.primaryQueueShard.Kind)
+	}
+
+	accountID := uuid.Nil
+	if sp.AccountID != nil {
+		accountID = *sp.AccountID
+	}
+
+	keys := []string{
+		q.primaryQueueShard.RedisClient.kg.BacklogSet(b.BacklogID),
+		q.primaryQueueShard.RedisClient.kg.ShadowPartitionSet(sp.PartitionID),
+		q.primaryQueueShard.RedisClient.kg.GlobalShadowPartitionSet(),
+		q.primaryQueueShard.RedisClient.kg.GlobalAccountShadowPartitions(),
+		q.primaryQueueShard.RedisClient.kg.AccountShadowPartitions(accountID),
+		q.primaryQueueShard.RedisClient.kg.PartitionQueueSet(enums.PartitionTypeDefault, sp.PartitionID, ""),
+	}
+	args, err := StrSlice([]any{
+		b.BacklogID,
+		sp.PartitionID,
+		accountID,
+	})
+	if err != nil {
+		return fmt.Errorf("could not serialize args: %w", err)
+	}
+
+	status, err := scripts["queue/backlogRefill"].Exec(
+		redis_telemetry.WithScriptName(ctx, "backlogRefill"),
+		q.primaryQueueShard.RedisClient.unshardedRc,
+		keys,
+		args,
+	).AsInt64()
+	if err != nil {
+		return fmt.Errorf("error refilling backlog: %w", err)
+	}
+	switch status {
+	case 0:
+		return nil
+	default:
+		return fmt.Errorf("unknown response refilling backlog: %v (%T)", status, status)
+	}
+}
+
+func (q *queue) ShadowPartitionPeek(ctx context.Context, sp *QueueShadowPartition, until time.Time, limit int64) ([]*QueueBacklog, error) {
+	ctx = redis_telemetry.WithScope(redis_telemetry.WithOpName(ctx, "ShadowPartitionPeek"), redis_telemetry.ScopeQueue)
+
+	if q.primaryQueueShard.Kind != string(enums.QueueShardKindRedis) {
+		return nil, fmt.Errorf("unsupported queue shard kind for partitionPeek: %s", q.primaryQueueShard.Kind)
+	}
+
+	if limit > ShadowPartitionPeekMaxBacklogs {
+		return nil, ErrShadowPartitionPeekMaxExceedsLimits
+	}
+	if limit <= 0 {
+		limit = ShadowPartitionPeekMaxBacklogs
+	}
+
+	ms := until.UnixMilli()
+
+	args, err := StrSlice([]any{
+		ms,
+		limit,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	rc := q.primaryQueueShard.RedisClient
+
+	shadowPartitionKey := rc.kg.ShadowPartitionSet(sp.PartitionID)
+
+	peekRet, err := scripts["queue/shadowPartitionPeek"].Exec(
+		redis_telemetry.WithScriptName(ctx, "shadowPartitionPeek"),
+		rc.Client(),
+		[]string{
+			shadowPartitionKey,
+			rc.kg.BacklogMeta(),
+		},
+		args,
+	).ToAny()
+
+	if err != nil {
+		return nil, fmt.Errorf("error peeking shadow partition backlogs: %w", err)
+	}
+	returnedSet, ok := peekRet.([]any)
+	if !ok {
+		return nil, fmt.Errorf("unknown return type from shadowPartitionPeek: %T", peekRet)
+	}
+
+	var potentiallyMissingBacklogs, allBacklogIDs []any
```

Comment:
do we still need all these for shadow partition processing?
what does it mean to have potentially missing backlogs?

Reply:
This is to clean up dangling backlog IDs in shadow partitions. I doubt this should happen but it’s probably still good to just clean them up, same for dangling queue items in backlogs themselves

Reply:
yeah, agree it's a good thing to have. kinda curious what will cause it.

Reply:
Honestly this should rarely, if ever, happen. For queue items in backlogs it's much more likely since the same item can be in more than a single backlog